src/sessionid.js

```diff
@@ -0,0 +1,89 @@
+import { INCREMENTAL_SNAPSHOT_EVENT_TYPE, MUTATION_SOURCE_TYPE } from './extensions/sessionrecording'
+import { SESSION_ID } from './posthog-persistence'
+import { sessionStore } from './storage'
+import { _ } from './utils'
+
+const SESSION_CHANGE_THRESHOLD = 30 * 60 * 1000 // 30 mins
+
+export class SessionIdManager {
+    constructor(config, persistence) {
+        this.persistence = persistence
+
+        if (config['persistence_name']) {
+            this.window_id_storage_key = 'ph_' + config['persistence_name'] + '_window_id'
+        } else {
+            this.window_id_storage_key = 'ph_' + config['token'] + '_window_id'
+        }
+    }
+
+    // Note: this tries to store the windowId in sessionStorage. SessionStorage is unique to the current window/tab,
+    // and persists page loads/reloads. So it's uniquely suited for storing the windowId. This function also respects
+    // when persistence is disabled (by user config) and when sessionStorage is not supported (it *should* be supported on all browsers),
+    // and in that case, it falls back to memory (which sadly, won't persist page loads)
+    _setWindowId(windowId) {
+        if (windowId !== this.windowId) {
+            this.windowId = windowId
+            if (!this.persistence.disabled && sessionStore.is_supported()) {
+                sessionStore.set(this.window_id_storage_key, windowId)
+            }
+        }
+    }
+
+    _getWindowId() {
+        if (this.windowId) {
+            return this.windowId
+        }
+        if (!this.persistence.disabled && sessionStore.is_supported()) {
+            return sessionStore.parse(this.window_id_storage_key)
+        }
+        return null
+    }
+
+    // Note: 'this.persistence.register' can be disabled in the config.
+    // In that case, this works by storing sessionId and the timestamp in memory.
+    _setSessionId(sessionId, timestamp) {
+        if (sessionId !== this.sessionId || timestamp !== this.timestamp) {
+            this.timestamp = timestamp
+            this.sessionId = sessionId
+            this.persistence.register({ [SESSION_ID]: [timestamp, sessionId] })
```

Comment:
So this is a bit tricky... 

We're using our persistence object here, and that decides between 4 types of persistence depending on user preference (localStorage, cookies, combination of local + cookies, memory). While the register function has an expiration on it, it looks like it would only work if cookie storage is chosen. Also, when cookies are selected, the persistence timeout actually applies to all information stored across the 'posthog cookie', so we would be expiring `distinct_id` etc.

Reply:
that makes sense!